client.go

```diff
@@ -1,48 +1,273 @@
+}
+
+// UploadWithPsk handles the uploading a file to AWS S3, into the bucket configured for this client, using a user-defined psk
+func (cli *S3) UploadWithPsk(ctx context.Context, req *UploadRequest, payload []byte, psk []byte) error {
+
+	listMultiInput := &s3.ListMultipartUploadsInput{
+		Bucket: &cli.bucketName,
+	}
+
+	listMultiOutput, err := cli.sdkClient.ListMultipartUploads(listMultiInput)
+	if err != nil {
+		log.Event(ctx, "error fetching multipart list", log.Error(err))
+		return err
+	}
+
+	var uploadID string
+	for _, upload := range listMultiOutput.Uploads {
+		if *upload.Key == req.UploadKey {
+			uploadID = *upload.UploadId
+		}
+	}
+
+	if len(uploadID) == 0 {
+
+		createMultiInput := &s3.CreateMultipartUploadInput{
+			Bucket:      &cli.bucketName,
+			Key:         &req.UploadKey,
+			ContentType: &req.Type,
+		}
+
+		createMultiOutput, err := cli.sdkClient.CreateMultipartUpload(createMultiInput)
+		if err != nil {
+			log.Event(ctx, "error creating multipart upload", log.Error(err))
+			return err
+		}
+
+		uploadID = *createMultiOutput.UploadId
+	}
+
+	rs := bytes.NewReader(payload)
+
+	n := int64(req.ChunkNumber)
+
+	uploadPartInput := &s3.UploadPartInput{
+		UploadId:   &uploadID,
+		Bucket:     &cli.bucketName,
+		Key:        &req.UploadKey,
+		Body:       rs,
+		PartNumber: &n,
+	}
+
+	if psk != nil {
+		_, err = cli.cryptoClient.UploadPartWithPSK(uploadPartInput, psk)
+		if err != nil {
+			log.Event(ctx, "error uploading with psk", log.Error(err))
+			return err
+		}
+	} else {
+		_, err = cli.sdkClient.UploadPart(uploadPartInput)
+		if err != nil {
+			log.Event(ctx, "error uploading part", log.Error(err))
+			return err
+		}
+	}
+
+	log.Event(ctx, "chunk accepted", log.Data{
+		"chunk_number": req.ChunkNumber,
+		"max_chunks":   req.TotalChunks,
+		"file_name":    req.FileName,
+	})
+
+	input := &s3.ListPartsInput{
+		Key:      &req.UploadKey,
+		Bucket:   &cli.bucketName,
+		UploadId: &uploadID,
+	}
+
+	output, err := cli.sdkClient.ListParts(input)
+	if err != nil {
+		log.Event(ctx, "error listing parts", log.Error(err))
+		return err
+	}
+
+	parts := output.Parts
+	if len(parts) == req.TotalChunks {
+		return cli.completeUpload(ctx, uploadID, req, parts)
+	}
+
+	return nil
+}
+
+// CheckUploaded check uploaded. Returns true only if the chunk corresponding to the provided chunkNumber has been uploaded.
+// If all the chunks have been uploaded, we complete the upload operation.
+func (cli *S3) CheckUploaded(ctx context.Context, req *UploadRequest) (bool, error) {
+
+	listMultiInput := &s3.ListMultipartUploadsInput{
+		Bucket: &cli.bucketName,
+	}
+
+	listMultiOutput, err := cli.sdkClient.ListMultipartUploads(listMultiInput)
+	if err != nil {
+		log.Event(ctx, "error fetching multipart upload list", log.Error(err))
+		return false, err
+	}
+
+	var uploadID string
+	for _, upload := range listMultiOutput.Uploads {
+		if *upload.Key == req.UploadKey {
+			uploadID = *upload.UploadId
```

Comment:
Does this need a `break` statement once an ID is assigned? Otherwise this could be overwritten.

Reply:
This is a good point! We could have concurrent upload operations against the same S3 key in the same bucket:
https://docs.aws.amazon.com/AmazonS3/latest/dev/mpuoverview.html#distributedmpupload

I will add a lock in the Upload function to prevent race conditions.

But in any case, the break will make the search more efficient.

Reply:
I will create a single mutex per client, but if we want to perform lots of 'get or create' uploads concurrently, we might want to have a mutex per s3 key.